flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/expressions/converter/ExpressionConverter.java

```diff
@@ -135,8 +138,24 @@ public RexNode visit(ValueLiteralExpression valueLiteral) {
 				return relBuilder.getRexBuilder().makeTimeLiteral(TimeString.fromCalendarFields(
 						valueAsCalendar(extractValue(valueLiteral, java.sql.Time.class))), 0);
 			case TIMESTAMP_WITHOUT_TIME_ZONE:
-				return relBuilder.getRexBuilder().makeTimestampLiteral(TimestampString.fromCalendarFields(
-						valueAsCalendar(extractValue(valueLiteral, java.sql.Timestamp.class))), 3);
+				TimestampType timestampType = (TimestampType) type;
+				Class<?> clazz = valueLiteral.getOutputDataType().getConversionClass();
+				LocalDateTime datetime = null;
+				if (clazz == LocalDateTime.class) {
+					datetime = extractValue(valueLiteral, LocalDateTime.class);
+				} else if (clazz == Timestamp.class) {
+					datetime = extractValue(valueLiteral, Timestamp.class).toLocalDateTime();
+				} else {
+					throw new TableException(String.format("Invalid literal of %s.", clazz.getCanonicalName()));
+				}
+				return relBuilder.getRexBuilder().makeTimestampLiteral(
+					new TimestampString(
```

Comment:
Please extract a static method `toTimestampString(LocalDateTime)` as previous PR said.

Reply:
There are two place have the `LocalDateTime` -> `TimestampString` logic: ExpressionConverter.java and ExpressionReducer.scala. I do not think it's a good idea to extract the logic to a function. For one thing, the logic is simple enough, wrap it adds complexity for readers/reviewers. For the other thing, no proper place to hold such a util now, add a util class for this is smelly.

Reply:
I would say it's worth adding methods.
First of all, this transformation is not simple. It requires 7 + method calls.
It's not a single constructor, it's easy to forget `withNanos`.
Besides, I don't think we will have nowhere to put down a method of reusing code.

Reply:
The only place which can hold such a util should be SqlDateTimeUtils, but it's in runtime module and don't recognize TimestampString( which in Calcite Core). So, let it go at that.